aries_cloudagent/protocols/issue_credential/v2_0/formats/vc_di/handler.py

```diff
@@ -0,0 +1,567 @@
+        in credential exchange messages. It looks up the correct format identifier and
+        encodes the data as a base64 attachment.
+
+        Args:
+            message_type (str): The message type for which to return the cred format.
+                Should be one of the message types defined in the message types file
+            data (dict): The data to include in the attach decorator
+        Returns:
+            CredFormatAttachment: Credential format and attachment data objects
+        """
+        return (
+            V20CredFormat(
+                attach_id=VCDICredFormatHandler.format.api,
+                format_=self.get_format_identifier(message_type),
+            ),
+            AttachDecorator.data_base64(data, ident=VCDICredFormatHandler.format.api),
+        )
+
+    async def _match_sent_cred_def_id(self, tag_query: Mapping[str, str]) -> str:
+        """Return most recent matching id of cred def that agent sent to ledger."""
+
+        async with self.profile.session() as session:
+            storage = session.inject(BaseStorage)
+            found = await storage.find_all_records(
+                type_filter=CRED_DEF_SENT_RECORD_TYPE, tag_query=tag_query
+            )
+        if not found:
+            raise V20CredFormatError(
+                f"Issuer has no operable cred def for proposal spec {tag_query}"
+            )
+        return max(found, key=lambda r: int(r.tags["epoch"])).tags["cred_def_id"]
+
+    async def create_proposal(
+        self, cred_ex_record: V20CredExRecord, proposal_data: Mapping[str, str]
+    ) -> Tuple[V20CredFormat, AttachDecorator]:
+        """Create vc_di credential proposal."""
+        if proposal_data is None:
+            proposal_data = {}
+
+        return self.get_format_data(CRED_20_PROPOSAL, proposal_data)
+
+    async def receive_proposal(
+        self, cred_ex_record: V20CredExRecord, cred_proposal_message: V20CredProposal
+    ) -> None:
+        """Receive vcdi credential proposal.
+
+        No custom handling is required for this step.
+        """
+
+    async def create_offer(
+        self, cred_proposal_message: V20CredProposal
+    ) -> CredFormatAttachment:
+        """Create vcdi credential offer."""
+
+        issuer = AnonCredsIssuer(self.profile)
+        ledger = self.profile.inject(BaseLedger)
+        cache = self.profile.inject_or(BaseCache)
+
+        async with self.profile.session() as session:
+            wallet = session.inject(BaseWallet)
+            public_did_info = await wallet.get_public_did()
+            public_did = public_did_info.did
+
+        cred_def_id = await issuer.match_created_credential_definitions(
+            **cred_proposal_message.attachment(VCDICredFormatHandler.format)
+        )
+
+        async def _create():
+            offer_str = await issuer.create_credential_offer(cred_def_id)
+            return json.loads(offer_str)
+
+        multitenant_mgr = self.profile.inject_or(BaseMultitenantManager)
```

Comment:
I feel like this ledger code should be cleaned up - either use the ledger registry in the anoncreds module, or move the functionality into the ledger class.  Not necessary for this PR (lets get it working and then schedule some cleanup tasks) but please add a `TODO` comment here.